File: apps/dsa-web/src/api/essayRadar.ts
import apiClient from './index';
import { toCamelCase } from './utils';
import type { EssayAnalysis, EssayAnalysisList, EssayAudioAnalysisCapability, EssayAudioAnalysisTask, EssayAudioBatchTask, EssayAudioDownloadProgress, EssayAudioFileList, EssayAudioTranscript, EssayCountBackfillResponse, EssayDailyReportList, EssayDashboard, EssayDeepInsights, EssayHistoricalBacklog, EssayInsights, EssayStatus, EssayWordCloud, EssayWorkerStatus, ResearchNoteDetail } from '../types/essayRadar';
import { cachedQuery } from './requestCache';

export type EssayFilters = {
  days?: number;
  query?: string;
  queryScope?: 'title' | 'full';
  sentiment?: string;
  category?: string;
  tag?: string;
  stock?: string;
  minImportance?: number;
  analysisStatus?: string;
  knownTotal?: number;
  page?: number;
  pageSize?: number;
};

export const essayRadarApi = {
  status: async (days = 30): Promise<EssayStatus> => {
    return cachedQuery(`essay:status:${days}`, async () => {
      const response = await apiClient.get('/api/v1/essay-radar/status', { params: { days } });
      return toCamelCase<EssayStatus>(response.data);
    }, { freshMs: 8_000, staleMs: 60_000 });
  },
  dashboard: async (days = 30): Promise<EssayDashboard> => {
    return cachedQuery(`essay:dashboard:${days}`, async () => {
      const response = await apiClient.get('/api/v1/essay-radar/dashboard', { params: { days } });
      return toCamelCase<EssayDashboard>(response.data);
    }, { freshMs: 15_000, staleMs: 120_000 });
  },
  insights: async (days = 30, trendDays = 14): Promise<EssayInsights> => {
    return cachedQuery(`essay:insights:${days}:${trendDays}`, async () => {
      const response = await apiClient.get('/api/v1/essay-radar/insights', { params: { days, trend_days: trendDays } });
      return toCamelCase<EssayInsights>(response.data);
    }, { freshMs: 15_000, staleMs: 120_000 });
  },
  deepInsights: async (params: {
    days?: number; trendDays?: number; horizon?: 'short' | 'medium' | 'long' | 'custom';
    startDate?: string; endDate?: string;
  } = {}): Promise<EssayDeepInsights> => {
    const key = `essay:deep:${JSON.stringify(params)}`;
    return cachedQuery(key, async () => {
      const response = await apiClient.get('/api/v1/essay-radar/deep-insights', {
        params: {
          days: params.days ?? 30,
          trend_days: params.trendDays ?? 14,
          horizon: params.horizon,
          start_date: params.startDate,
          end_date: params.endDate,
        },
        timeout: 120000,
      });
      return toCamelCase<EssayDeepInsights>(response.data);
    }, { freshMs: 30_000, staleMs: 600_000 });
  },
  interpretMarketImpact: async (payload: {
    tsCode: string; horizon: 'short' | 'medium' | 'long' | 'custom'; startDate?: string; endDate?: string;
  }): Promise<{
    generatedAt: string; model: string; tsCode: string;
    interpretation: { conclusion: string; evidence: string[]; limitations: string[]; nextChecks: string[] };
  }> => {
    const response = await apiClient.post('/api/v1/essay-radar/deep-insights/market-interpretation', {
      ts_code: payload.tsCode,
      horizon: payload.horizon,
      start_date: payload.startDate,
      end_date: payload.endDate,
    }, { timeout: 120000 });
    return toCamelCase(response.data);
  },
  wordCloud: async (period: 'day' | 'week' | 'month', kind: 'stocks' | 'tags' | 'themes' = 'stocks'): Promise<EssayWordCloud> => {
    return cachedQuery(`essay:word-cloud:${period}:${kind}`, async () => {
      const response = await apiClient.get('/api/v1/essay-radar/word-cloud', { params: { period, kind } });
      return toCamelCase<EssayWordCloud>(response.data);
    }, { freshMs: 15_000, staleMs: 120_000 });
  },
  dailyReports: async (): Promise<EssayDailyReportList> => {
    return cachedQuery('essay:daily-reports', async () => {
      const response = await apiClient.get('/api/v1/essay-radar/daily-reports', { params: { limit: 30 } });
      return toCamelCase<EssayDailyReportList>(response.data);
    }, { freshMs: 30_000, staleMs: 300_000 });
  },
  runDailyReports: async (): Promise<unknown> => {
    const response = await apiClient.post('/api/v1/essay-radar/daily-reports/run', { force: false }, { timeout: 300000 });
    return toCamelCase(response.data);
  },
  list: async (filters: EssayFilters): Promise<EssayAnalysisList> => {
    const response = await apiClient.get('/api/v1/essay-radar/feed', {
      params: {
        days: filters.days ?? 0,
        query: filters.query || undefined,
        query_scope: filters.queryScope ?? 'full',
        analysis_status: filters.analysisStatus || undefined,
        known_total: filters.knownTotal,
        sentiment: filters.sentiment || undefined,
        category: filters.category || undefined,
        tag: filters.tag || undefined,
        stock: filters.stock || undefined,
        min_importance: filters.minImportance ? filters.minImportance : undefined,
        page: filters.page ?? 1,
        page_size: filters.pageSize ?? 20,
      },
    });
    return toCamelCase<EssayAnalysisList>(response.data);
  },
  exportFeed: async (filters: EssayFilters): Promise<Blob> => {
    const response = await apiClient.get('/api/v1/essay-radar/feed/export', {
      params: {
        days: filters.days ?? 0,
        query: filters.query || undefined,
        query_scope: filters.queryScope ?? 'full',
        analysis_status: filters.analysisStatus || undefined,
        sentiment: filters.sentiment || undefined,
        category: filters.category || undefined,
        tag: filters.tag || undefined,
        stock: filters.stock || undefined,
        min_importance: filters.minImportance ? filters.minImportance : undefined,
      },
      responseType: 'blob',
      timeout: 300000,
    });
    return response.data as Blob;
  },
  exportSelected: async (topicIds: string[]): Promise<Blob> => {
    const response = await apiClient.post('/api/v1/essay-radar/feed/export-selected', {
      topic_ids: topicIds,
    }, {
      responseType: 'blob',
      timeout: 300000,
    });
    return response.data as Blob;
  },
  audioFiles: async (filters: Pick<EssayFilters, 'days' | 'query' | 'page' | 'pageSize'>): Promise<EssayAudioFileList> => {
    const response = await apiClient.get('/api/v1/financial-data/research-notes/audio-files', {
      params: {
        days: filters.days ?? 0,
        query: filters.query || undefined,
        page: filters.page ?? 1,
        page_size: filters.pageSize ?? 20,
      },
    });
    return toCamelCase<EssayAudioFileList>(response.data);
  },
  downloadSelectedAudio: async (items: Array<{ topicId: string; fileId: string }>): Promise<Blob> => {
    const response = await apiClient.post('/api/v1/financial-data/research-notes/audio-files/batch-download', {
      items: items.map((item) => ({ topic_id: item.topicId, file_id: item.fileId })),
    }, {
      responseType: 'blob',
      timeout: 600000,
    });
    return response.data as Blob;
  },
  startAudioBatchTask: async (items: Array<{ topicId: string; fileId: string }>): Promise<EssayAudioBatchTask> => {
    const response = await apiClient.post('/api/v1/financial-data/research-notes/audio-files/batch-download-tasks', {
      items: items.map((item) => ({ topic_id: item.topicId, file_id: item.fileId })),
    });
    return toCamelCase<EssayAudioBatchTask>(response.data);
  },
  audioBatchTask: async (taskId: string): Promise<EssayAudioBatchTask> => {
    const response = await apiClient.get(`/api/v1/financial-data/research-notes/audio-files/batch-download-tasks/${encodeURIComponent(taskId)}`);
    return toCamelCase<EssayAudioBatchTask>(response.data);
  },
  audioBatchTasks: async (): Promise<{ items: EssayAudioBatchTask[]; total: number }> => {
    const response = await apiClient.get('/api/v1/financial-data/research-notes/audio-files/batch-download-tasks', { params: { limit: 20 } });
    return toCamelCase(response.data);
  },
  downloadAudioBatchTask: async (
    taskId: string,
    onProgress?: (progress: EssayAudioDownloadProgress) => void,
  ): Promise<Blob> => {
    const response = await apiClient.get(`/api/v1/financial-data/research-notes/audio-files/batch-download-tasks/${encodeURIComponent(taskId)}/download`, {
      responseType: 'blob',
      timeout: 600000,
      onDownloadProgress: (event) => {
        const total = event.total && event.total > 0 ? event.total : undefined;
        onProgress?.({
          loaded: event.loaded,
          total,
          percent: total ? Math.min(100, Math.round((event.loaded / total) * 100)) : undefined,
        });
      },
    });
    return response.data as Blob;
  },
  audioAnalysisCapability: async (): Promise<EssayAudioAnalysisCapability> => {
    const response = await apiClient.get('/api/v1/financial-data/research-notes/audio-analysis/capability');
    return toCamelCase<EssayAudioAnalysisCapability>(response.data);
  },
  startAudioAnalysisTask: async (
    items: Array<{ topicId: string; fileId: string }>,
    options: { title?: string; focus?: string; hotwords?: string[]; speakerCount?: number; generateMemo?: boolean } = {},
  ): Promise<EssayAudioAnalysisTask> => {
    const response = await apiClient.post('/api/v1/financial-data/research-notes/audio-analysis/tasks', {
      items: items.map((item) => ({ topic_id: item.topicId, file_id: item.fileId })),
      title: options.title || undefined,
      focus: options.focus || undefined,
      hotwords: options.hotwords || [],
      speaker_count: options.speakerCount || undefined,
      generate_memo: options.generateMemo ?? true,
    });
    return toCamelCase<EssayAudioAnalysisTask>(response.data);
  },
  audioAnalysisTask: async (taskId: string): Promise<EssayAudioAnalysisTask> => {
    const response = await apiClient.get(`/api/v1/financial-data/research-notes/audio-analysis/tasks/${encodeURIComponent(taskId)}`);
    return toCamelCase<EssayAudioAnalysisTask>(response.data);
  },
  audioAnalysisTasks: async (): Promise<{ items: EssayAudioAnalysisTask[]; total: number }> => {
    const response = await apiClient.get('/api/v1/financial-data/research-notes/audio-analysis/tasks', { params: { limit: 20 } });
    return toCamelCase(response.data);
  },
  retryAudioAnalysisTask: async (taskId: string): Promise<EssayAudioAnalysisTask> => {
    const response = await apiClient.post(`/api/v1/financial-data/research-notes/audio-analysis/tasks/${encodeURIComponent(taskId)}/retry`);
    return toCamelCase<EssayAudioAnalysisTask>(response.data);
  },
  audioAnalysisTranscript: async (taskId: string, fileId: string): Promise<EssayAudioTranscript> => {
    const response = await apiClient.get(`/api/v1/financial-data/research-notes/audio-analysis/tasks/${encodeURIComponent(taskId)}/transcripts/${encodeURIComponent(fileId)}`);
    return toCamelCase<EssayAudioTranscript>(response.data);
  },
  downloadAudioTranscript: async (taskId: string, fileId: string): Promise<Blob> => {
    const response = await apiClient.get(
      `/api/v1/financial-data/research-notes/audio-analysis/tasks/${encodeURIComponent(taskId)}/transcripts/${encodeURIComponent(fileId)}/download`,
      { responseType: 'blob', timeout: 300000 },
    );
    return response.data as Blob;
  },
  downloadSelectedAudioTranscripts: async (items: Array<{ topicId: string; fileId: string }>): Promise<Blob> => {
    const response = await apiClient.post('/api/v1/financial-data/research-notes/audio-files/transcripts/batch-download', {
      items: items.map((item) => ({ topic_id: item.topicId, file_id: item.fileId })),
    }, { responseType: 'blob', timeout: 300000 });
    return response.data as Blob;
  },
  downloadAudioAnalysis: async (taskId: string, format: 'zip' | 'md' | 'docx' | 'json'): Promise<Blob> => {
    const response = await apiClient.get(`/api/v1/financial-data/research-notes/audio-analysis/tasks/${encodeURIComponent(taskId)}/download`, {
      params: { format }, responseType: 'blob', timeout: 600000,
    });
    return response.data as Blob;
  },
  detail: async (topicId: string): Promise<EssayAnalysis> => {
    const response = await apiClient.get(`/api/v1/essay-radar/analyses/${encodeURIComponent(topicId)}`);
    return toCamelCase<EssayAnalysis>(response.data);
  },
  note: async (topicId: string): Promise<ResearchNoteDetail> => {
    const response = await apiClient.get(`/api/v1/financial-data/research-notes/${encodeURIComponent(topicId)}`);
    return toCamelCase<ResearchNoteDetail>(response.data);
  },
  backfill: async (days = 30): Promise<unknown> => {
    const response = await apiClient.post('/api/v1/essay-radar/backfill', { days }, { timeout: 180000 });
    return toCamelCase(response.data);
  },
  historicalBacklog: async (): Promise<EssayHistoricalBacklog> => {
    const response = await apiClient.get('/api/v1/essay-radar/historical-backlog');
    // camelcase-keys treats a trailing unit after a number as an acronym
    // (notes_24h -> notes24H). Normalize these three public fields explicitly
    // so the dashboard cannot silently show zero for factual activity data.
    const converted = toCamelCase<EssayHistoricalBacklog & {
      notes24H?: number;
      notes7D?: number;
      notes30D?: number;
    }>(response.data);
    return {
      ...converted,
      notes24h: converted.notes24H ?? converted.notes24h ?? 0,
      notes7d: converted.notes7D ?? converted.notes7d ?? 0,
      notes30d: converted.notes30D ?? converted.notes30d ?? 0,
    };
  },
  backfillCount: async (count: number, order: 'newest' | 'oldest'): Promise<EssayCountBackfillResponse> => {
    const response = await apiClient.post(
      '/api/v1/essay-radar/backfill-count',
      { count, order },
      { timeout: 180000 },
    );
    return toCamelCase<EssayCountBackfillResponse>(response.data);
  },
  startWorker: async (): Promise<EssayWorkerStatus> => {
    const response = await apiClient.post('/api/v1/essay-radar/worker/start');
    return toCamelCase<EssayWorkerStatus>(response.data);
  },
  stopWorker: async (): Promise<EssayWorkerStatus> => {
    const response = await apiClient.post('/api/v1/essay-radar/worker/stop');
    return toCamelCase<EssayWorkerStatus>(response.data);
  },
  retryFailed: async (): Promise<unknown> => {
    const response = await apiClient.post('/api/v1/essay-radar/retry-failed', { start_worker: true });
    return toCamelCase(response.data);
  },
  syncMcp: async (): Promise<unknown> => {
    const response = await apiClient.post('/api/v1/financial-data/zsxq/sync', undefined, { timeout: 120000 });
    return toCamelCase(response.data);
  },
  backfillMcpHistory: async (years: 1 | 2): Promise<unknown> => {
    const response = await apiClient.post('/api/v1/financial-data/zsxq/history/backfill', { years });
    return toCamelCase(response.data);
  },
  startMcpWorker: async (): Promise<unknown> => {
    const response = await apiClient.post('/api/v1/financial-data/zsxq/sync/worker/start');
    return toCamelCase(response.data);
  },
  stopMcpWorker: async (): Promise<unknown> => {
    const response = await apiClient.post('/api/v1/financial-data/zsxq/sync/worker/stop');
    return toCamelCase(response.data);
  },
};

File: api/v1/endpoints/essay_radar.py
# -*- coding: utf-8 -*-
"""DeepSeek research-note tagging, queue and dashboard endpoints."""

from __future__ import annotations

from typing import Optional

from fastapi import APIRouter, HTTPException, Query

from api.v1.schemas.essay_radar import (
    EssayBackfillRequest,
    EssayCountBackfillRequest,
    EssayDailyReportRunRequest,
    EssayRetryRequest,
)
from src.services.essay_analysis_service import EssayAnalysisError, EssayAnalysisService, EssayDailyReportService
from src.services.essay_daily_report_worker import EssayDailyReportWorker
from src.services.essay_analysis_worker import EssayAnalysisWorker
from src.services.zsxq_mcp_sync_service import ZsxqMcpSyncWorker

router = APIRouter()


def _service_error(exc: Exception, status_code: int = 400) -> HTTPException:
    return HTTPException(
        status_code=status_code,
        detail={"error": "essay_analysis_error", "message": str(exc)[:500]},
    )


@router.get("/status", summary="读取近 30 天分析进度与后台任务状态")
def status(days: int = Query(30, ge=1, le=3650)):
    return {
        "progress": EssayAnalysisService().progress(days=days),
        "worker": EssayAnalysisWorker.get_instance().status(),
        "mcp_sync": ZsxqMcpSyncWorker.get_instance().status(),
        "daily_report_worker": EssayDailyReportWorker.get_instance().status(),
    }


@router.post("/backfill", summary="将指定天数内纪要加入分析队列")
def backfill(request: EssayBackfillRequest):
    service = EssayAnalysisService()
    result = service.enqueue_recent(days=request.days)
    try:
        worker = EssayAnalysisWorker.get_instance().start()
    except EssayAnalysisError as exc:
        raise _service_error(exc, 503)
    return {"queue": result, "worker": worker}


@router.get("/historical-backlog", summary="读取全部已入库小作文的 AI 分析与未入队数量")
def historical_backlog():
    return EssayAnalysisService().historical_backlog()


@router.post("/backfill-count", summary="按篇数选择尚未入队的历史小作文进行 AI 分析")
def backfill_count(request: EssayCountBackfillRequest):
    service = EssayAnalysisService()
    result = service.enqueue_unqueued(count=request.count, order=request.order)
    try:
        # This entry point must honor the requested count exactly. New MCP
        # records are already queued at ingest, so no 30-day bootstrap is needed.
        worker = EssayAnalysisWorker.get_instance().start(bootstrap_recent=False)
    except EssayAnalysisError as exc:
        raise _service_error(exc, 503)
    return {"queue": result, "backlog": service.historical_backlog(), "worker": worker}


@router.post("/worker/start", summary="启动实时 DeepSeek 分析任务")
def start_worker():
    try:
        return EssayAnalysisWorker.get_instance().start()
    except EssayAnalysisError as exc:
        raise _service_error(exc, 503)


@router.post("/worker/stop", summary="停止实时 DeepSeek 分析任务")
def stop_worker():
    return EssayAnalysisWorker.get_instance().stop()


@router.post("/retry-failed", summary="重置失败任务并可立即启动后台任务")
def retry_failed(request: EssayRetryRequest):
    retried = EssayAnalysisService().repo.retry_failed()
    try:
        worker = EssayAnalysisWorker.get_instance().start() if request.start_worker else None
    except EssayAnalysisError as exc:
        raise _service_error(exc, 503)
    return {"retried": retried, "worker": worker}


@router.get("/dashboard", summary="获取小作文标签和股票热度看板")
def dashboard(
    days: int = Query(30, ge=1, le=3650),
    top_n: int = Query(12, ge=3, le=30),
):
    return EssayAnalysisService().dashboard(days=days, top_n=top_n)


@router.get("/insights", summary="获取小作文趋势、证据质量、模型共识与关注股洞察")
def insights(
    days: int = Query(30, ge=1, le=3650),
    trend_days: int = Query(14, ge=7, le=90),
):
    return EssayAnalysisService().insights(days=days, trend_days=trend_days)


@router.get("/deep-insights", summary="获取小作文来源、主题、标的、催化风险多层洞察")
def deep_insights(
    days: int = Query(30, ge=7, le=3650),
    trend_days: int = Query(14, ge=7, le=90),
):
    return EssayAnalysisService().deep_insights(days=days, trend_days=trend_days)


@router.get("/word-cloud", summary="获取日、周、月股票/标签/主题词云")
def word_cloud(
    period: str = Query("day", pattern="^(day|week|month)$"),
    kind: str = Query("stocks", pattern="^(stocks|tags|themes)$"),
    anchor_date: Optional[str] = Query(None, pattern=r"^\d{4}-\d{2}-\d{2}$"),
    stock: Optional[str] = None,
    top_n: int = Query(80, ge=10, le=200),
):
    try:
        return EssayAnalysisService().word_cloud(period=period, kind=kind, anchor_date=anchor_date, stock=stock, top_n=top_n)
    except ValueError as exc:
        raise _service_error(exc, 400)


@router.get("/daily-reports", summary="读取各模型的前一日小作文日报")
def daily_reports(limit: int = Query(30, ge=1, le=366), model: Optional[str] = None):
    return EssayDailyReportService().list(limit=limit, model=model)


@router.post("/daily-reports/run", summary="立即生成指定日期的多模型小作文日报")
def run_daily_reports(request: EssayDailyReportRunRequest):
    return EssayDailyReportWorker.get_instance().run_now(request.report_date, request.force)


@router.post("/daily-reports/worker/start", summary="启动每日多模型报告任务")
def start_daily_report_worker():
    return EssayDailyReportWorker.get_instance().start()


@router.post("/daily-reports/worker/stop", summary="停止每日多模型报告任务")
def stop_daily_report_worker():
    return EssayDailyReportWorker.get_instance().stop()


@router.get("/analyses", summary="筛选和分页读取小作文分析结果")
def list_analyses(
    days: int = Query(30, ge=1, le=3650),
    query: Optional[str] = None,
    status: Optional[str] = None,
    sentiment: Optional[str] = None,
    category: Optional[str] = None,
    tag: Optional[str] = None,
    stock: Optional[str] = None,
    min_importance: Optional[int] = Query(None, ge=0, le=100),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
):
    return EssayAnalysisService().list_analyses(
        days=days,
        query=query,
        status=status,
        sentiment=sentiment,
        category=category,
        tag=tag,
        stock=stock,
        min_importance=min_importance,
        page=page,
        page_size=page_size,
    )


@router.get("/feed", summary="检索全部已入库小作文，可选关联 AI 分析结果")
def list_feed(
    days: int = Query(0, ge=0, le=3650),
    query: Optional[str] = Query(None, max_length=200),
    analysis_status: Optional[str] = Query(
        None,
        pattern="^(completed|uncompleted|not_queued|pending|processing|failed)$",
    ),
    sentiment: Optional[str] = None,
    category: Optional[str] = None,
    tag: Optional[str] = None,
    stock: Optional[str] = None,
    min_importance: Optional[int] = Query(None, ge=0, le=100),
    known_total: Optional[int] = Query(None, ge=0),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
):
    return EssayAnalysisService().list_feed(
        days=days,
        query=query,
        analysis_status=analysis_status,
        sentiment=sentiment,
        category=category,
        tag=tag,
        stock=stock,
        min_importance=min_importance,
        known_total=known_total,
        page=page,
        page_size=page_size,
    )


@router.get("/analyses/{topic_id}", summary="读取单篇小作文分析详情")
def get_analysis(topic_id: str):
    try:
        return EssayAnalysisService().get_analysis(topic_id)
    except KeyError as exc:
        raise _service_error(exc, 404)
